dali/imgcodec/decoders/nvjpeg.cc

```diff
@@ -0,0 +1,169 @@
+      }
+    }
+  } else if (strcmp(name, "host_memory_padding") == 0) {
+    host_memory_padding_ = any_cast<size_t>(value);
+
+    if (host_memory_padding_ > 0) {
+      for (auto thread_id : tp_->GetThreadIds()) {
+        nvjpeg_memory::AddHostBuffer(thread_id, host_memory_padding_);
+        nvjpeg_memory::AddHostBuffer(thread_id, host_memory_padding_);
+      }
+    }
+  }
+}
+
+any NvJpegDecoderInstance::GetParam(const char *name) const {
+  if (strcmp(name, "device_memory_padding") == 0) {
+    return device_memory_padding_;
+  } else if (strcmp(name, "host_memory_padding") == 0) {
+    return host_memory_padding_;
+  } else {
+    DALI_FAIL("Unrecognized param name");
+  }
+}
+
+DecodeResult NvJpegDecoderInstance::DecodeImplTask(int thread_idx,
+                                                   cudaStream_t stream,
+                                                   SampleView<GPUBackend> out,
+                                                   ImageSource *in,
+                                                   DecodeParams opts,
+                                                   const ROI &roi) {
+  DecodingContext ctx = DecodingContext {
+    .resources = resources_[thread_idx],
+  };
+  CUDA_CALL(nvjpegDecodeParamsCreate(nvjpeg_handle_, &ctx.params));
+  CUDA_CALL(nvjpegDecodeParamsSetOutputFormat(ctx.params, GetFormat(opts.format)));
+  CUDA_CALL(nvjpegDecodeParamsSetAllowCMYK(ctx.params, true));
+
+  try {
+    ParseJpeg(*in, opts, ctx);
+    DecodeJpeg(*in, out.mutable_data<uint8_t>(), opts, ctx);
+  } catch (...) {
+    return {false, std::make_exception_ptr(std::current_exception())};
+  }
+
+  return {true, nullptr};
+}
+
+void NvJpegDecoderInstance::ParseJpeg(ImageSource& in, DecodeParams opts, DecodingContext& ctx) {
+  int widths[NVJPEG_MAX_COMPONENT], heights[NVJPEG_MAX_COMPONENT], c;
+  nvjpegChromaSubsampling_t subsampling;
+  cudaResultCheck(nvjpegGetImageInfo(nvjpeg_handle_, in.RawData<unsigned char>(), in.Size(), &c,
+                                     &subsampling, widths, heights));
+
+  ctx.shape = {heights[0], widths[0], c};
+}
+
+void NvJpegDecoderInstance::DecodeJpeg(ImageSource& in, uint8_t *out, DecodeParams opts,
+                                       DecodingContext &ctx) {
+  auto& decoder = ctx.resources.decoder_data.decoder;
+  auto& state = ctx.resources.decoder_data.state;
+  auto& jpeg_stream = ctx.resources.jpeg_stream;
+  auto& stream = ctx.resources.stream;
+  auto& decode_event = ctx.resources.decode_event;
+  auto& device_buffer = ctx.resources.device_buffer;
+
+  cudaResultCheck(nvjpegStateAttachPinnedBuffer(state, ctx.resources.pinned_buffer));
+  cudaResultCheck(nvjpegJpegStreamParse(nvjpeg_handle_, in.RawData<unsigned char>(), in.Size(),
+                                        false, false, ctx.resources.jpeg_stream));
+  cudaResultCheck(nvjpegDecodeJpegHost(nvjpeg_handle_, decoder, state, ctx.params, jpeg_stream));
```

Comment:
Doesn't CUDA_CALL work here?
```suggestion
  CUDA_CALL(nvjpegStateAttachPinnedBuffer(state, ctx.resources.pinned_buffer));
  CUDA_CALL(nvjpegJpegStreamParse(nvjpeg_handle_, in.RawData<unsigned char>(), in.Size(),
                                        false, false, ctx.resources.jpeg_stream));
  CUDA_CALL(nvjpegDecodeJpegHost(nvjpeg_handle_, decoder, state, ctx.params, jpeg_stream));
```

Reply:
Oops, I haven't noticed that `CUDA_CALL` triggers `cudaResultCheck`. Will fix